// gbrain/src/core/embedding.ts
/**
 * Embedding Service - Custom GLM Implementation
 * Uses GLM embedding-2 (1024 dimensions) via direct HTTPS call
 * Bypasses OpenAI SDK to avoid compatibility issues
 */

const MODEL = 'text-embedding-3-large';
const DIMENSIONS = 1024;
const MAX_CHARS = 8000;
const MAX_RETRIES = 5;
const BASE_DELAY_MS = 4000;
const MAX_DELAY_MS = 120000;
const BATCH_SIZE = 100;

function getGLMKey(): string {
  const envKey = process.env.GBRAIN_GLM_KEY || process.env.GLM_KEY || '';
  if (envKey) return envKey;
  try {
    const fs = require('fs');
    const envContent = fs.readFileSync('/home/lighthouse/.env', 'utf8');
    const match = envContent.match(/glmkey=(.+)/);
    if (match) return match[1].trim();
  } catch {}
  return process.env.OPENAI_API_KEY || '';
}

const GLM_KEY = getGLMKey();
const GLM_URL = 'https://open.bigmodel.cn/api/paas/v4/embeddings';

export async function embed(text: string): Promise<Float32Array> {
  const truncated = text.slice(0, MAX_CHARS);
  const result = await embedBatch([truncated]);
  return result[0];
}

export async function embedBatch(texts: string[]): Promise<Float32Array[]> {
  const truncated = texts.map(t => t.slice(0, MAX_CHARS));
  const results: Float32Array[] = [];
  for (let i = 0; i < truncated.length; i += BATCH_SIZE) {
    const batch = truncated.slice(i, i + BATCH_SIZE);
    const batchResults = await embedBatchWithRetry(batch);
    results.push(...batchResults);
  }
  return results;
}

async function embedBatchWithRetry(texts: string[]): Promise<Float32Array[]> {
  for (let attempt = 0; attempt < MAX_RETRIES; attempt++) {
    try {
      const response = await callGLMEmbedding(texts);
      return response;
    } catch (err: any) {
      if (attempt === MAX_RETRIES - 1) throw err;
      const delay = Math.min(BASE_DELAY_MS * 2 ** attempt, MAX_DELAY_MS);
      await new Promise(r => setTimeout(r, delay));
    }
  }
  return [];
}

async function callGLMEmbedding(texts: string[]): Promise<Float32Array[]> {
  const glmPayload = { model: 'embedding-2', input: texts };
  const response = await fetch(GLM_URL, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${GLM_KEY}`,
    },
    body: JSON.stringify(glmPayload),
  });
  if (!response.ok) {
    const text = await response.text();
    throw new Error(`GLM API error ${response.status}: ${text}`);
  }
  const data = await response.json();
  const embeddings = data.data || [];
  const sorted = embeddings.sort((a: any, b: any) => a.index - b.index);
  return sorted.map((d: any) => new Float32Array(d.embedding));
}

export function normalize(embedding: Float32Array, text: string): Float32Array {
  if (text.length < 256) {
    const factor = 10;
    const result = new Float32Array(embedding.length);
    for (let i = 0; i < embedding.length; i++) {
      result[i] = embedding[i] * factor;
    }
    return result;
  }
  return embedding;
}

export { MODEL as EMBEDDING_MODEL, DIMENSIONS as EMBEDDING_DIMENSIONS };
